files/galaxy/dynamic_job_rules/production/total_perspective_vortex/users.yml

```diff
@@ -1,4 +1,82 @@
 users:
+    default:
+      rules:
+        - id: slurm_training_rule
+          if: |
+            training_job_matches_specs = False  # return value
+            training_rules_apply = False
+            try:
+              if any([r for r in user.all_roles() if (not r.deleted and r.name.startswith('training'))]):
+                training_rules_apply = True
+              elif 'training' in [hta.user_tname for hta in job.history.tags]:
+                training_rules_apply = True
+            except Exception as e:
+              log.debug('--tpv log debug: Exception checking training status of job')
+              log.debug(str(e))
+
+            if training_rules_apply:
+              include_tags = []
+              exclude_tags = ['pulsar', 'training-exempt']
+              slurm_preferred_for_training = pulsar_score is not None and pulsar_score < pulsar_score_slurm_threshold and not tag_values_match(entity, include_tags=[], exclude_tags=['pulsar-training-large', 'training-exempt'])
+              training_job_matches_specs = slurm_preferred_for_training or helpers.tag_values_match(entity, include_tags, exclude_tags)
+            training_job_matches_specs
+          max_cores: 2
+          max_mem: max_cores * 3.8
+          context:
+            partition: training
+          scheduling:
+            require:
+              - slurm
+              - training
+        - id: pulsar_training_rule
+          if: |
+            training_job_matches_specs = False  # return value
+            training_rules_apply = False
+            try:
+              if any([r for r in user.all_roles() if (not r.deleted and r.name.startswith('training'))]):
+                training_rules_apply = True
+              elif 'training' in [hta.user_tname for hta in job.history.tags]:
+                training_rules_apply = True
+            except Exception as e:
+              log.debug('--tpv log debug: Exception checking training status of job')
+              log.debug(str(e))
+              
+            if training_rules_apply:
+              include_tags = ['pulsar']
+              exclude_tags = ['pulsar-training-large', 'training-exempt']
+              slurm_preferred_for_training = pulsar_score is not None and pulsar_score < pulsar_score_slurm_threshold
+              training_job_matches_specs = (not slurm_preferred_for_training) and helpers.tag_values_match(entity, include_tags, exclude_tags)
```

Comment:
Since only this line is different between slurm and pulsar, maybe it would make sense to move this code to a seperate file that we just import. A possible future enhancement. I imagine more scenarios like this will pile up.

Reply:
I agree - most of the code in the 3 blocks is identical.